infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/PreviewUrlCommandProvisioner.java

```diff
@@ -0,0 +1,115 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+package org.eclipse.che.workspace.infrastructure.kubernetes.provision;
+
+import static org.eclipse.che.api.core.model.workspace.config.Command.PREVIEW_URL_ATTRIBUTE;
+
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.ServicePort;
+import io.fabric8.kubernetes.api.model.extensions.HTTPIngressPath;
+import io.fabric8.kubernetes.api.model.extensions.Ingress;
+import io.fabric8.kubernetes.api.model.extensions.IngressBackend;
+import io.fabric8.kubernetes.api.model.extensions.IngressRule;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.inject.Singleton;
+import org.eclipse.che.api.workspace.server.model.impl.CommandImpl;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.Services;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Updates commands with proper Preview URL attribute.
+ *
+ * <p>Goes through all {@link CommandImpl}s in given {@link KubernetesEnvironment} and for ones that
+ * have defined Preview URL tries to find matching {@link Service} and {@link Ingress} from given
+ * {@link KubernetesNamespace}. When found, it composes full URL and set it as attribute of a
+ * command with key {@link
+ * org.eclipse.che.api.core.model.workspace.config.Command#PREVIEW_URL_ATTRIBUTE}.
+ *
+ * @param <E> type of the environment
+ */
+@Singleton
+public class PreviewUrlCommandProvisioner<E extends KubernetesEnvironment> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PreviewUrlCommandProvisioner.class);
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    injectsPreviewUrlToCommands(env, namespace);
+  }
+
+  /**
+   * Go through all commands, find matching service and exposed host. Then construct full preview
+   * url from this data and set it as Command's parameter under `previewUrl` key.
+   *
+   * @param env environment to get commands
+   * @param namespace current kubernetes namespace where we're looking for services and ingresses
+   */
+  private void injectsPreviewUrlToCommands(E env, KubernetesNamespace namespace)
+      throws InfrastructureException {
+    if (env.getCommands() == null) {
+      return;
+    }
+    for (CommandImpl command :
+        env.getCommands()
+            .stream()
+            .filter(c -> c.getPreviewUrl() != null)
+            .collect(Collectors.toList())) {
+      Optional<Service> foundService =
+          Services.findServiceWithPort(
+              namespace.services().get(), command.getPreviewUrl().getPort());
```

Comment:
You fetch the same services list for each command with previewURL, since it's K8s API call via http we could fetch it once and reuse. The same about Ingresses list https://github.com/eclipse/che/pull/14713/files#diff-25437c9721ce7375785a2f31969cef82R102.

Reply:
done